Upgrade (still private) ghostwriter

Zac-HD opened the pull request:
In this stage, we're adding functions to write tests for several new properties, as well as documentation and support for unittest-style tests.

Part of #2344, follows #2449 - in particular this PR handles strategies with syntactically-invalid reprs, but doesn't yet include "approval tests" or golden-master tests as I found the `approvaltests` package pretty immature (close to a decade old, still at version 0.2.x, minimal-to-no documentation, etc.).

TODO in future PRs: golden master testing, the CLI, and maybe some heuristics to automatically guess which properties to test / which functions in a module to test.

Zac-HD commented:
@HypothesisWorks/hypothesis-python-contributors - a quick review would be appreciated, because I'd really like to have the PR after this one (CLI and publicise it) done in time for the SciPy sprints on July 11th 😀

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -14,8 +14,37 @@
-See https://github.com/HypothesisWorks/hypothesis/pull/2344 for progress.
+Writing tests with Hypothesis frees you from the tedium of deciding on and
+writing out specific inputs to test.  Now, the ``hypothesis.ghostwriter``
+module can even write your source code for you!
```
this sounds like it's writing the source code for the main code rather than for the tests. Maybe "write your tests for you"?

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -14,8 +14,37 @@
+(if Black is installed, the ghostwriter uses it automatically).
+
+Legal questions?  While the ghostwriter fragments and logic is under the
+MPL-2.0 license like the rest of Hypothesis, the *output* from the ghostwriter
```
Would it be worth adding a comment to this tune to the top of generated files?

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -86,16 +122,22 @@ def _strategy_for(param: inspect.Parameter) -> Union[st.SearchStrategy, InferTyp
-def _get_strategies(*funcs: Callable) -> Dict[str, st.SearchStrategy]:
+def _get_strategies(
+    *funcs: Callable, passon: bool = False
+) -> Dict[str, st.SearchStrategy]:
```
I can't really tell what this function is doing. In particular I don't know what the `passon` parameter is for. Could it get a docstring?

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -104,23 +146,53 @@ def _get_strategies(*funcs: Callable) -> Dict[str, st.SearchStrategy]:
+    r = (
+        repr(strategy)
+        .replace("sampled_from([False, True])", "booleans()")
+        .replace("just(None)", "none()")
```
Some of these I wonder if we should replace in the strategy reprs themselves. e.g. it seems not unreasonable to have these special cases for `just` and `sampled_from` in their `__repr__` definition.

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -156,6 +228,12 @@ def _make_test(
 
+    if style == "unittest":
+        imports.add("unittest")
+        body = "class Test{}{}(unittest.TestCase):\n".format(
```
You don't have to do this, but I wonder if you this would be better handled using the pretty printer module we have vendored? It's good for managing indentation and that sort of thing.

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -173,14 +251,191 @@ def _make_test(
+            try:
+                re.compile(pattern=pattern, flags=flags)
+            except sre_constants.error:
+                assume(False)
```
This should be `reject()` (it is in the actual code)

DRMacIver commented on hypothesis-python/src/hypothesis/ghostwriter.py:
```diff
@@ -173,14 +251,191 @@ def _make_test(
+                    *zip(params[1:], params[1:]),
+                ]
+            )
+        test_body += "value{} = {}.{}({})\n".format(i, f.__module__, f.__name__, args)
```
Yeah I think I actually really want this code to use the pretty printer module or something like it to write its results. This kind of string concatenation code is fairly unpleasant to read.

Zac-HD commented:
@DRMacIver - main updates since your review:

1. I tried implementing this in terms of the pretty-printer, but the structure turns into a complete mess of context managers at best - assembling parts and working out how far to indent them later is still cleaner.  I've added a few helper functions to cut down on the complicated formatting chains though, which definitely helps.
2. I've moved it to `extra.ghostwriter`, so that we can have a hard-dependency on `black`.  This also requires that the Ghostwriter be 3.6+, so I've gone over to f-strings for simple formatting too.

DRMacIver commented:
> I tried implementing this in terms of the pretty-printer, but the structure turns into a complete mess of context managers at best - assembling parts and working out how far to indent them later is still cleaner. I've added a few helper functions to cut down on the complicated formatting chains though, which definitely helps.

Fair enough!

> I also added golden master tests in cf6c7ec, but backed them out again because most of them ended up varying slightly by minor version 😬

Huh. Minor version of what? Black? Do you know why that is?

Zac-HD commented:
Minor version of Python! The two I remember is that the deprecated `encoding` argument of json.loads() was recently removed, and that some of the regex constants are defined in the sre_constants module but in some versions set their module attribute to re.

DRMacIver commented:
> Minor version of Python! The two I remember is that the deprecated encoding argument of json.loads() was recently removed, and that some of the regex constants are defined in the sre_constants module but in some versions set their module attribute to re.

Hmm. Right. Maybe we want the golden master tests for this to only run on a very specific Python version? We could put them in their own test repo and make a dedicated task for it that runs on a fixed version - easy enough to update that and just run the usual "update all the masters" task.

Zac-HD commented:
- I've added a skip to `test_find_large_union_list`, because per #2497 it's failing on all my PRs.
- Turns out that the version-specific outputs are mostly fixable with some simple string-replacements, so I've done that.  It's not elegant, but it's commented and it does work...

Zac-HD commented:
This has gotten too big; I'm going to split it up for easier review.
